File: app/src/ultimate/java/com/clockmods/ultimate/AntiBurnPreferences.java
package com.clockmods.ultimate;

import android.content.Context;
import android.content.SharedPreferences;

/** Global anti-burn settings shared by every Ultimate destination. */
public final class AntiBurnPreferences {
    public static final String PREFERENCES_NAME = "clockmods_anti_burn";
    public static final boolean DEFAULT_ENABLED = false;
    public static final int DEFAULT_PERIOD_MINUTES = 10;
    public static final float DEFAULT_AMPLITUDE_DP = 4f;
    public static final boolean DEFAULT_AUTO_DIM = true;
    private static final String KEY_ENABLED = "enabled";
    private static final String KEY_PERIOD = "period_minutes";
    private static final String KEY_AMPLITUDE = "amplitude_dp";
    private static final String KEY_AUTO_DIM = "auto_dim";
    private final SharedPreferences preferences;

    public AntiBurnPreferences(Context context) {
        if (context == null) throw new IllegalArgumentException("context must not be null");
        preferences = context.getApplicationContext().getSharedPreferences(
                PREFERENCES_NAME, Context.MODE_PRIVATE);
    }

    public boolean isEnabled() { return preferences.getBoolean(KEY_ENABLED, DEFAULT_ENABLED); }
    public void setEnabled(boolean value) { preferences.edit().putBoolean(KEY_ENABLED, value).apply(); }
    public int getPeriodMinutes() { return normalizePeriod(preferences.getInt(KEY_PERIOD,
            DEFAULT_PERIOD_MINUTES)); }
    public void setPeriodMinutes(int value) { preferences.edit().putInt(KEY_PERIOD,
            normalizePeriod(value)).apply(); }
    public float getAmplitudeDp() { return normalizeAmplitude(preferences.getFloat(KEY_AMPLITUDE,
            DEFAULT_AMPLITUDE_DP)); }
    public void setAmplitudeDp(float value) { preferences.edit().putFloat(KEY_AMPLITUDE,
            normalizeAmplitude(value)).apply(); }
    public boolean isAutoDim() { return preferences.getBoolean(KEY_AUTO_DIM, DEFAULT_AUTO_DIM); }
    public void setAutoDim(boolean value) { preferences.edit().putBoolean(KEY_AUTO_DIM, value).apply(); }

    public static int normalizePeriod(int value) {
        if (value == 1 || value == 30 || value == 60) return value;
        return 10;
    }

    public static float normalizeAmplitude(float value) {
        if (Float.isNaN(value) || Float.isInfinite(value)) return DEFAULT_AMPLITUDE_DP;
        return Math.max(0f, Math.min(12f, value));
    }
}

File: app/src/ultimate/java/com/clockmods/ultimate/AntiBurnPreferences.kt
package com.clockmods.ultimate

import android.content.Context
import android.content.SharedPreferences

/** Global anti-burn settings shared by every Ultimate destination. */
class AntiBurnPreferences {
    private val preferences: SharedPreferences

    constructor(context: Context?) {
        preferences = requireNotNull(context) {
            "context must not be null"
        }.applicationContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
    }

    /** Constructor retained for JVM tests and isolated preference stores. */
    internal constructor(preferences: SharedPreferences) {
        this.preferences = preferences
    }

    fun isEnabled(): Boolean = preferences.getBoolean(KEY_ENABLED, DEFAULT_ENABLED)

    fun setEnabled(value: Boolean) {
        preferences.edit().putBoolean(KEY_ENABLED, value).apply()
    }

    fun getPeriodMinutes(): Int = normalizePeriod(
        preferences.getInt(KEY_PERIOD, DEFAULT_PERIOD_MINUTES),
    )

    fun setPeriodMinutes(value: Int) {
        preferences.edit().putInt(KEY_PERIOD, normalizePeriod(value)).apply()
    }

    fun getAmplitudeDp(): Float = normalizeAmplitude(
        preferences.getFloat(KEY_AMPLITUDE, DEFAULT_AMPLITUDE_DP),
    )

    fun setAmplitudeDp(value: Float) {
        preferences.edit().putFloat(KEY_AMPLITUDE, normalizeAmplitude(value)).apply()
    }

    fun isAutoDim(): Boolean = preferences.getBoolean(KEY_AUTO_DIM, DEFAULT_AUTO_DIM)

    fun setAutoDim(value: Boolean) {
        preferences.edit().putBoolean(KEY_AUTO_DIM, value).apply()
    }

    fun restoreDefaults() {
        preferences.edit().clear().apply()
    }

    companion object {
        const val PREFERENCES_NAME = "clockmods_anti_burn"
        const val DEFAULT_ENABLED = false
        const val DEFAULT_PERIOD_MINUTES = 10
        const val DEFAULT_AMPLITUDE_DP = 4f
        const val DEFAULT_AUTO_DIM = true

        private const val KEY_ENABLED = "enabled"
        private const val KEY_PERIOD = "period_minutes"
        private const val KEY_AMPLITUDE = "amplitude_dp"
        private const val KEY_AUTO_DIM = "auto_dim"

        @JvmStatic
        fun normalizePeriod(value: Int): Int = when (value) {
            1, 30, 60 -> value
            else -> DEFAULT_PERIOD_MINUTES
        }

        @JvmStatic
        fun normalizeAmplitude(value: Float): Float =
            if (!value.isFinite()) DEFAULT_AMPLITUDE_DP else value.coerceIn(0f, 12f)
    }
}
